From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
Date: Thu, 21 Mar 2019 09:11:55 +0100
Subject: [PATCH] {For LLVMLinux} arch/x86/include/asm/arch_hweight: fix clang
 build

arch/x86/boot/string.o is compiled using "-m16", which triggers the
following error with clang:

    In file included from arch/x86/boot/string.c:16:
    In file included from include/linux/kernel.h:12:
    In file included from include/linux/bitops.h:19:
    In file included from arch/x86/include/asm/bitops.h:508:
    arch/x86/include/asm/arch_hweight.h:51:15: error: invalid input size
    for constraint 'D'
                             : REG_IN (w));
                                       ^

'D' means "%rdi" only on __x86_64__.
---
 arch/x86/include/asm/arch_hweight.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/x86/include/asm/arch_hweight.h b/arch/x86/include/asm/arch_hweight.h
index fc0693569f7a..8676bd77b862 100644
--- a/arch/x86/include/asm/arch_hweight.h
+++ b/arch/x86/include/asm/arch_hweight.h
@@ -4,7 +4,7 @@
 
 #include <asm/cpufeatures.h>
 
-#ifdef CONFIG_64BIT
+#if defined(CONFIG_64BIT) && defined(__x86_64__)
 #define REG_IN "D"
 #define REG_OUT "a"
 #else
@@ -35,7 +35,7 @@ static inline unsigned int __arch_hweight8(unsigned int w)
 	return __arch_hweight32(w & 0xff);
 }
 
-#ifdef CONFIG_X86_32
+#if defined(CONFIG_X86_32) || !defined(__x86_64__)
 static inline unsigned long __arch_hweight64(__u64 w)
 {
 	return  __arch_hweight32((u32)w) +
-- 
